src/modules/fw_pos_control/performance_model/PerformanceModel.cpp

```diff
@@ -0,0 +1,225 @@
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+
+/**
+ * @file PerformanceModel.cpp
+ * Performance model.
+ */
+
+#include <geo/geo.h>
+#include <px4_platform_common/events.h>
+#include "PerformanceModel.h"
+
+// air density of standard athmosphere at 5000m above mean sea level [kg/m^3]
+static constexpr float kAirDensityStandardAtmos5000Amsl = 0.7363f;
+
+// air density of standard athmosphere at 1000m above mean sea level [kg/m^3]
+static constexpr float kAirDensityStandardAtmos1000Amsl = 1.112f;
+
+// [.] minimum ratio between the actual vehicle weight and the vehicle nominal weight (weight at which the performance limits are derived)
+static constexpr float kMinWeightRatio = 0.5f;
+
+// [.] maximum ratio between the actual vehicle weight and the vehicle nominal weight (weight at which the performance limits are derived)
+static constexpr float kMaxWeightRatio = 2.0f;
+
+// climbrate defining the service ceiling, used to compensate max climbrate based on air density
+static constexpr float kClimbrateMin = 0.5f; // [m/s]
+
+PerformanceModel::PerformanceModel(): ModuleParams(nullptr)
+{
+	updateParams();
+}
+float PerformanceModel::getWeightRatio() const
+{
+	float weight_factor = 1.0f;
+
+	if (_param_weight_base.get() > FLT_EPSILON && _param_weight_gross.get() > FLT_EPSILON) {
+		weight_factor = math::constrain(_param_weight_gross.get() / _param_weight_base.get(), kMinWeightRatio,
+						kMaxWeightRatio);
+	}
+
+	return weight_factor;
+}
+float PerformanceModel::getMaximumClimbRate(float air_density) const
+{
+
+	float climbrate_max = _param_fw_t_clmb_max.get();
+
+	const float density_min = _param_density_min.get();
+
+	if (density_min < kAirDensityStandardAtmos1000Amsl
+	    && density_min > kAirDensityStandardAtmos5000Amsl) {
+		const float density_gradient = math::max((_param_fw_t_clmb_max.get() - kClimbrateMin) /
+					       (CONSTANTS_AIR_DENSITY_SEA_LEVEL_15C -
+						density_min), 0.0f);
+		const float delta_rho = air_density - CONSTANTS_AIR_DENSITY_SEA_LEVEL_15C;
```

Comment:
Should we also constrain the air_density input? To values between CONSTANTS_AIR_DENSITY_SEA_LEVEL_15C and kAirDensityStandardAtmos5000Amsl

Reply:
done https://github.com/PX4/PX4-Autopilot/blob/b898d8529e797902be12d5d52282a64184fa390c/src/lib/fw_performance_model/PerformanceModel.cpp#L90